enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+see-also:
+  - "https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md"
+  - "https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md"
+  - "https://github.com/openshift/enhancements/blob/master/enhancements/etcd/disaster-recovery-with-ceo.md"
+  - "https://github.com/openshift/machine-config-operator/blob/master/docs/etcd-quorum-guard.md"
+  - "https://github.com/openshift/cluster-kube-scheduler-operator"
+  - "https://github.com/openshift/cluster-kube-controller-manager-operator"
+  - "https://github.com/openshift/cluster-openshift-controller-manager-operator"
+  - "https://github.com/openshift/cluster-kube-apiserver-operator"
+  - "https://github.com/openshift/cluster-openshift-apiserver-operator"
+---
+
+# Single entity Control Plane
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+For clarity in this doc we set the definition for "Control Plane" as "The collection of stateless and stateful processes which enable a Kubernetes cluster to meet minimum operational requirements". This includes: kube-apiserver, kube-controller-manager, kube-scheduler, kubelet and Etcd.
+
+This proposal outlines a solution for declaratively managing as a single entity the compute resources that host the OCP Control Plane components. It introduces scaling and self-healing capabilities for this compute resources while honouring inviolable Etcd expectations and with out disrupting the lifecycle of Control plane components.
+
+## Motivation
+
+The Control Plane is the most critical and sensitive entity of a running cluster. Today OCP Control Plane instances are "pets" and therefore fragile. There are multiple scenarios where adjusting the compute capacity which is backing the Control Plane components might be desirable either for resizing or repairing. 
+Currently there is nothing that automates or eases this task. The steps for the Control Plane to be resized in any manner or to recover from a tolerable failure are completely manual. Different teams are following different SOP documents scattered around with manual steps resulting in lost of information, confusion and extra efforts for engineers and users.
+
+### Goals
+
+- To have a declarative mechanism to manage the Control Plane as a single entity.
+- To support declarative safe horizontal scaling towards an odd number replicas for the control plane compute resources.
+- To support declarative safe vertical scaling for the control plane compute resources.
+- To support declarative safe self-healing and replacement of Control Plane compute resources when the unhealthy resources are below n/2.
```

Comment:
If etcd is below n/2, it is in a read-only state. How will self healing work if etcd is frozen? Why wouldn't we begin self-healing the moment that _any_ of the resources are unhealthy?

Reply:
Yes, It would self-heal as far as the number of healthy members is > n/2. 
Let me rephrase this.